expr.go

```diff
@@ -10,29 +10,44 @@ import (
 )
 
 // EvalPredicate - Evaluate a given expression as a boolean predicate against a set of fields and tags
-func EvalPredicate(se stateful.Expression, now time.Time, fields models.Fields, tags models.Tags) (bool, error) {
-	vars, err := mergeFieldsAndTags(now, fields, tags)
+func EvalPredicate(se stateful.Expression, scopePool stateful.ScopePool, now time.Time, fields models.Fields, tags models.Tags) (bool, error) {
+	vars := scopePool.Get()
+	defer scopePool.Put(vars)
+	err := fillScope(vars, scopePool.ReferenceVariables(), now, fields, tags)
 	if err != nil {
 		return false, err
 	}
+
 	b, err := se.EvalBool(vars)
 	if err != nil {
 		return false, err
 	}
 	return b, nil
 }
 
-func mergeFieldsAndTags(now time.Time, fields models.Fields, tags models.Tags) (*tick.Scope, error) {
-	scope := tick.NewScope()
-	scope.Set("time", now.Local())
-	for k, v := range fields {
-		if _, ok := tags[k]; ok {
-			return nil, fmt.Errorf("cannot have field and tags with same name %q", k)
+// fillScope - given a scope and reference variables, we fill the exact variables from the now, fields and tags.
+func fillScope(vars *tick.Scope, referenceVariables []string, now time.Time, fields models.Fields, tags models.Tags) error {
+	for _, refVariableName := range referenceVariables {
+		if refVariableName == "time" {
+			vars.Set("time", now.Local())
```

Comment:
Should we `break` after this line? That way we don't accidentally set a field value named `time`.